### DISCUSSION TITLE
IMU Pre-Integration in Python

### DISCUSSION THREAD

**Discussion by chrisflesher** [symforce-org/symforce#402] [Q&A]
Hello, today is the first day I was looking at SymForce. I pip installed symforce v0.9.0 with python3.8 from PyPi, that worked. The first thing I tried was to attempt IMU pre-integration but was unable to import the objects needed...

Tried:
```
from symforce.slam.imu_factor import ImuFactor
from symforce.slam.imu_preintegration import ImuPreintegrator
import symforce.slam.imu_preintegration.imu_factor
import symforce.slam.imu_preintegration.imu_preintegrator
```
These don't seem to be installed?

Maybe this is related to issue #323?

**Comment by chrisflesher** [comment]
`README.md` suggest building from source, this built successfully but still can't seem to import `ImuFactor` or `ImuPreintegrator`.

Is the answer to use the code in `manifold_symbolic`? If so it would be great to see some example code or unit tests for this, I didn't find any in the repo.

Thanks!

**Comment by aaron-skydio** [comment]
So, there are a couple components involved here.  It depends on what you're interested in - are you just interested in numerically performing IMU preintegration, or are you interested in playing with the symbolic expressions?

`manifold_symbolic` is indeed the definition of the symbolic components of IMU preintegration.  There aren't really examples of using this beyond the code that uses it to generate the numerical `ImuFactor`, that code is in `symforce/slam/imu_preintegration/generate.py`.

For numerically performing IMU preintegration, `ImuFactor` and `ImuPreintegrator` are the right objects, but they're defined in C++.  You can see examples of how to use them in C++ in `slam_imu_preintegration_test.cc`.

`ImuPreintegrator` and friends are also wrapped using `pybind11` so they're available in Python.  If you want to use them in Python to do numerical preintegration, you can find them under `symforce.cc_sym`, documented [here](https://symforce.org/api/symforce.cc_sym.html#cc_sym.ImuPreintegrator).  We don't currently have a wrapper for `ImuFactor` though.
